### PR TITLE
New encoder

### PR DESCRIPTION
**Pull request by Minchan-Kim**
I created a new encoder which uses attention mechanism instead of cropping (It also uses skip connections inside CNN block). I checked that it works (without error). Beware that feature_extractor.py (which contains the legacy feature extractor NetHackNet) does not exists now. The new encoder is in feature.py and its name is NetHackEncoder (not NethackEncoder, which was originally in that file).

### CODE REVIEW

File: feature.py [hunk 1]
```diff
@@ -67,57 +67,75 @@ def forward(self, x: th.Tensor) -> th.Tensor:
...
-            CBA(32, 64, (3, 3), padding=0),
-            nn.Flatten()
+            # -- [K x 79 x 21]
+            Residual(in_channels[0], out_channels[0]),
+            # -- [A x 27 x 7]
+            Residual(in_channels[1], out_channels[1], padding=(1, 0), stride=3),
+            # -- [A x 9 x 3]
+            CBA(in_channels[2], out_channels[2], padding=(0, 1), stride=3),
+            # -- [A x 27]
+            nn.Flatten(start_dim=2)
```
Comment (yycho0108) [inline feature.py:106]: Didn't know about `start_dim` argument in `nn.Flatten`!

File: feature.py [hunk 2]
```diff
@@ -67,57 +67,75 @@ def forward(self, x: th.Tensor) -> th.Tensor:
...
+        c_in: int, 
+        c_out: int, 
+        padding: Union[int, Tuple[int, int]] = 1, 
+        stride: Union[int, Tuple[int, int]] = 1
+    ):
+        super().__init__()
+        self.conv1 = CBA(c_in, c_out, padding=padding, stride=stride)
+        self.conv2 = CBA(c_out, c_out, lrelu_eps=None)
+        self.residual = nn.Sequential()
+        if c_in != c_out or stride != 1:
```
Comment (yycho0108) [inline feature.py:81]: Why check for `if c_in != c_out or stride != 1` rather than `if use_residual`?

File: feature.py [hunk 3]
```diff
@@ -67,57 +67,75 @@ def forward(self, x: th.Tensor) -> th.Tensor:
...
 
     def forward(self, x: th.Tensor) -> th.Tensor:
         """
         Args:
-            x: int16 tensor of shape (..., 21, 79),
-               in range (0, 5976).
+            x: tensor of shape (..., embedding_dim, 21, 79).
         Returns:
-            y: (..., D) feature embedding of `x`.
+            y: (..., A, 27) feature embedding of `x`.
```
Comment (yycho0108) [inline feature.py:114]: What is `A`? Is it just some arbitrary dimension? Or does it have some meaning?

**Review by yycho0108 — APPROVED** [review]
Overall LGTM. Great job on implementing a whole new feature encoder and attention module!

File: feature.py [hunk 4]
```diff
@@ -67,57 +67,75 @@ def forward(self, x: th.Tensor) -> th.Tensor:
...
+        self, 
+        c_in: int, 
+        c_out: int, 
+        padding: Union[int, Tuple[int, int]] = 1, 
+        stride: Union[int, Tuple[int, int]] = 1
+    ):
+        super().__init__()
+        self.conv1 = CBA(c_in, c_out, padding=padding, stride=stride)
+        self.conv2 = CBA(c_out, c_out, lrelu_eps=None)
+        self.residual = nn.Sequential()
```
Comment (yycho0108) [inline feature.py:80]: Does just using `nn.Sequential()` actually work, when the below `if` block evaluated to `False`?

File: feature.py (hunk 3 above)
Comment (Minchan-Kim) [reply feature.py:114]: A stands for Attention. Dimension 'A' is used for query.

File: feature.py (hunk 2 above)
Comment (Minchan-Kim) [reply feature.py:81]: I thought about introducing a new variable like use_residual, but somehow I preferred not to create a new one...

File: feature.py (hunk 4 above)
Comment (Minchan-Kim) [reply feature.py:80]: As far as I know (and it should be, since the first residual block uses empty nn.Sequential()), it works like an identity layer.

**Pull request merged**